### PR TITLE
feat(ui-shell): add side nav rail

### LINKED ISSUE #2996: [Header, SideNav] - Development - UX for rails pattern
- [ ]  Develop UX for header with rails
    - sidenav collapses into visible icons
    - expands on hover, but on `HeaderMenuButton` click expands indefinitely
- [ ] Update component readme if needed (will also help update readme on website)
- [ ] Add any necessary new documentation to website (style page for new classes, component status page, other)
- [ ] Add story for `SideNav Rails w/ Header`

cc: @designertyler  for design specs/review

### PR DESCRIPTION
**Pull request by aledavila**
Closes #2996 

#### Changelog

**New**

- Side Nav Rails variation
- added a class to hide menu button

Story variation for rails is removed from storybook.

Known issues: *Will be addressed in v10.5
- Side nav sub-menu needs to collapse when rail is closed (Bug)
- Side nav should open when sub menu item is focused (Enhancement)

### CODE REVIEW

File: packages/react/src/components/UIShell/HeaderMenuButton.js
Comment (asudoh) [inline packages/react/src/components/UIShell/HeaderMenuButton.js]: Probably you meant just `!isCollapsible`?

**Review by asudoh — COMMENTED** [review]
Just wanted to drop a note though I know this is WIP. Looking forward to seeing the ready-for-review version @aledavila!

**Review by tw15egan — COMMENTED** [review]
Question for the @designertyler // @aagonzales : What should happen on collapse and a submenu item is active?

Screen Shot 2019-06-26 at 9 03 05 AM


Ah nvm, looks like it should be handled like this:

Screen Shot 2019-06-26 at 9 05 18 AM

File: packages/react/src/components/UIShell/SideNav.js
Comment (tw15egan) [inline packages/react/src/components/UIShell/SideNav.js]: Could we change this to:

```js
onMouseEnter={() => handleMouseOver(true)}
onMouseLeave={() => handleMouseOver(false)}
```
and then on line 44: 

```js
  const handleMouseOver = value => {
    if (isRail) {
      setExpandedState(value);
    }
  };
```

File: packages/react/src/components/UIShell/HeaderMenuButton.js
Comment (aledavila) [reply packages/react/src/components/UIShell/HeaderMenuButton.js]: Yes thanks

**Comment by aledavila** [comment]
@vpicone 
Mentioned that side nav needs to be refactored in order to work as intended. 
Will open a new issue for that and to be able to handle the side nav sub menu closing on collapse.

**Comment by tw15egan** [comment]
Yeah, looks like `SideNav` will need to pass down something, or just toggle it directly when the rail is closed

**Review by emyarod — COMMENTED** [review]
looks like the sassdoc needs to be updated again for CI

has the collapsed sidenav submenu issue been created?

**Comment by tw15egan** [comment]
@emyarod Looks like the issue was created here: https://github.com/carbon-design-system/carbon/issues/3223

**Review by tw15egan — APPROVED** [review]
Looks good to me, nice work! 🎉

**Comment by joshblack** [comment]
@aledavila anything else you need to help get this reviewed/merged in?

**Comment by aledavila** [comment]
@joshblack no it was blocked by Josefina's pr but thats now merged in so have to go in and fix the merge conflicts. It just needs design review after that

**Comment by joshblack** [comment]
cc @designertyler do you have time to review after the merge conflicts get addressed?

**Comment by aledavila** [comment]
@designertyler its ready for review.

**Review by dakahn — APPROVED** [review]

**Comment by designertyler** [comment]
Visuals are spot on. I wasn't able to check the selected state when the rail is collapsed. Is that part of Vince's work?
 
- [ ] The rail should expand when a nav item is focused. 
image

**Review by tw15egan — COMMENTED** [review]
Where is the `packages/icons/meta.json` coming from? @joshblack does this file need to be ignored or just removed from the PR

**Comment by aledavila** [comment]
@tw15egan Not sure. It happened in a local merge when there was conflicts. I can remove it @joshblack ?

**Comment by joshblack** [comment]
Haha, good catch. It was renamed to `build-info.json` so  `meta.json` can safely be removed!

**Comment by tw15egan** [comment]
@designertyler are you good with merging this as-is, and working on the two known issues in `10.5`? The rails version will not be displayed in the storybook until `10.5`

**Review by designertyler — APPROVED** [review]
Yes, @aledavila let me know and that sounds fine.

**Pull request merged**
